Add Google Spreadsheet Example (#1551)
wso2/docs-ei: en/micro-integrator/docs/references/connectors/google-spreadsheet-connector/google-spreadsheet-connector-example.md

# Google Spreadsheet Connector Example

The Google Sheets API lets users to read and modify any aspect of a spreadsheet. The WSO2 Google Spreadsheet Connector allows you to access the Google Spreadsheet [API Version v4](https://developers.google.com/sheets/api/guides/concepts) through WSO2 EI. It allows users to read/write any aspect of the spreadsheet via the spreadsheets collection.

It has been provided the ability to do Spreadsheet operations and Spreadsheet Data operations. 

## What you'll build

This example explains how to use Google Spreadsheet Connector to create a Google spreadsheet, write data to it and get them read. Further, it explains how the data in the spreadsheet can be edited. 

It will have three HTTP API resources, which are `insert`, `read` and `edit`. 

> * `/insert `: The user sends the request payload which includes the name of the spreadsheet should be, what are the sheet names and what data should be inserted to which sheet and the which range of cells.  This request is sent to WSO2 EI by invoking the Spreadsheet API. It creates a spreadsheet with specified data in the specified cell rage. 

Reviewer: "to which sheet and the which range of cells" some grammar issue.

Author: Changed

> * `/insert `: The user sends the request payload which includes the name of the spreadsheet should be, what are the sheet names and what data should be inserted to which sheet and the which range of cells.  This request is sent to WSO2 EI by invoking the Spreadsheet API. It creates a spreadsheet with specified data in the specified cell rage. 

Reviewer 2: ```suggestion
* `/insert `: The user sends the request payload, which includes the name of the spreadsheet, the sheet names, and what data should be inserted to which sheet and which range of cells.  This request is sent to WSO2 EI by invoking the Spreadsheet API. It creates a spreadsheet with specified data in the specified cell range. 
```

Author: Done

> * `/read `: The user sends the request payload, which includes the spreadsheet Id which should be obtained from calling the `insert` API resource, and the range of the cell range to be read. 

Reviewer 2: ```suggestion
* `/read `: The user sends the request payload, which includes the spreadsheet Id that should be obtained from calling the `insert` API resource, and the range of the cell range to be read. 
```

Author: Done

> * `/edit `: The user sends the request payload which includes the spreadsheet Id which should be obtained from calling the `insert` API resource, and the data to be edited that includes values and the range. 

Reviewer 2: ```suggestion
* `/edit `: The user sends the request payload, which includes the spreadsheet Id that should be obtained from calling the `insert` API resource, and the data to be edited that includes values and the range. 
```

Author: Done

If you do not want to configure this yourself, you can simply [get the project](#get-the-project) and run it.

## Configure the connector in WSO2 Integration Studio

> Follow these steps to set up the ESB Solution Project and the Connector Exporter Project. 

Reviewer 2: ```suggestion
Follow these steps to set up the Integration Project and the Connector Exporter Project. 
```

Author: Done

{!references/connectors/importing-connector-to-integration-studio.md!} 

## Creating the Integration Logic

1. Follow these steps to [Configure Google Sheets API](get-credentials-for-google-spreadsheet.md) and obtain the Client Id, Client Secret, Access Token and Refresh Token.  

2. Right click on the created Integration Project and select, -> **New** -> **Rest API** to create the REST API. 
    <img src="../../../../assets/img/connectors/adding-an-api.png" title="Adding a Rest API" width="800" alt="Adding a Rest API"/>

3. Provide the API name as `SpreadsheetAPI` and the API context as `/insert`.

4. First we will create the `/insert` resource. Right click on the API Resource and go to **Properties** view. We use a URL template called `/insert` as we have two API resources inside single API. The method will be `Post`. 
    <img src="../../../../assets/img/connectors/filecon-3.png" title="Adding the API resource." width="800" alt="Adding the API resource."/>

> 5. In this operation we are going to receive input from the user which are `properties`, `sheets`, `range` and `values`. 

Reviewer 2: ```suggestion
5. In this operation we are going to receive input from the user, which are `properties`, `sheets`, `range` and `values`. 
```

Author: Done

    - properties - It can provide the spreadsheet properties such as title of the spreadsheet. 
    - sheets - It can provide set of sheets to be created.
    - range - It provides the sheet name and the range that data need to be inserted. 
    - values - Data to be inserted.

> 6. The above four parameters are saved to a property group. Drag and drop the Property Group mediator onto the canvas in the design view and do as shown below. For further reference, you can read about the [Property Group mediator](https://ei.docs.wso2.com/en/latest/micro-integrator/references/mediators/property-Group-Mediator/). You can add set of properties as below. 

Reviewer 2: Let's make this a relative link.

Author: Done

    - clientId
    - clientSecret
    - refreshToken

    <img src="/assets/img/connectors/sheetcon3.png" title="init operation" width="800" alt="init operation"/>

10. Drag and drop **createSpreadsheet** operation to the Canvas next. Parameter values are defined in step 6 and 7 in the property group. 

    <img src="/assets/img/connectors/sheetcon4.png" title="Parameters" width="800" alt="Parameters"/>

11. The complete XML configuration for the `createSpreadsheet.xml` looks as below. 
```xml
    <?xml version="1.0" encoding="UTF-8"?>
    <sequence name="createSpreadsheet" trace="disable" xmlns="http://ws.apache.org/ns/synapse">
        <googlespreadsheet.init>
            <accessToken></accessToken>
            <apiUrl>https://sheets.googleapis.com/v4/spreadsheets</apiUrl>
            <clientId></clientId>
            <clientSecret></clientSecret>
            <refreshToken></refreshToken>
        </googlespreadsheet.init>
        <googlespreadsheet.createSpreadsheet>
            <properties>{$ctx:properties}</properties>
            <sheets>{$ctx:sheets}</sheets>
        </googlespreadsheet.createSpreadsheet>
    </sequence>
```

> 12. Next we need to create the `addData.xml` sequence as above. As explained in step 8, create a sequence by right clicking the Integration Project which has already been created. 

Reviewer 2: ```suggestion
12. Next we need to create the `addData.xml` sequence as above. As explained in step 8, create a sequence by right clicking the Integration Project that has already been created. 
```

Author: Done

            <outSequence/>
            <faultSequence/>
        </resource>
        <resource methods="POST" uri-template="/read">
            <inSequence>
                <sequence description="This sequence will read data of the spreadsheet. " key="readData"/>
                <respond/>
            </inSequence>
            <outSequence/>
            <faultSequence/>
        </resource>
        <resource methods="POST" uri-template="/edit">
            <inSequence>
                <sequence key="editSpreadsheet"/>
                <respond/>
            </inSequence>
            <outSequence/>
            <faultSequence/>
        </resource>
    </api>

```

{!references/connectors/exporting-artifacts.md!}

## Get the project

You can download the ZIP file and extract the contents to get the project code.

<a href="../../../../assets/attach/connectors/FileConnector.zip">
    <img src="../../../../assets/img/connectors/download-zip.png" width="200" alt="Download ZIP">
</a>

## Deployment

Follow these steps to deploy the exported CApp in the Enterprise Integrator Runtime. 

{!references/connectors/deploy-capp.md!}

## Testing

### Spreadsheet insert Operation

> 1. Invoke the SpreadsheetAPI with the following URL. Application such as [Postman](https://www.postman.com/) can be used to invoke the API. 

Reviewer 2: ```suggestion
Invoke the SpreadsheetAPI with the following URL. An application such as [Postman](https://www.postman.com/) can be used to invoke the API. 
```

Author: Done

      "properties":{
          "title": "Company"
        },
      "sheets":[
          {
            "properties":
            {
              "title": "Employees"
            }
          },
          {
            "properties":
            {
              "title": "Hector"
            }
          }
        ], 
        "range":"Employees!A1:C3",
        "values":[
            [
                "First Name",
                "Last Name",
                "Gender"
            ],
            [
                "John",
                "Doe",
                "Male"
            ],
            [
              "Leon",
                "Wins",
                "Female"
              ]
        ]
    }
    
  ```
**Expected Response**: 
You should get a success response as below, and the spreadsheet should be created in the given ID in the response with data inserted.  

```
  {
    "spreadsheetId": "1ddnO00fcjuLvEMCUORVjYQ4C0VLeAPNGmcvSvELHbPU",
    "updates": {
      "spreadsheetId": "1ddnO00fcjuLvEMCUORVjYQ4C0VLeAPNGmcvSvELHbPU",
      "updatedRange": "Employees!A1:C3",
      "updatedRows": 3,
      "updatedColumns": 3,
      "updatedCells": 9
    }
  }
```

### Spreadsheet Read Operation

> 1. Invoke the SpreadsheetAPI with the following URL. Application such as [Postman](https://www.postman.com/) can be used to invoke the API. Obtain the Spreadsheet ID from the step 1. 

Reviewer 2: ```suggestion
Invoke the SpreadsheetAPI with the following URL. Application such as [Postman](https://www.postman.com/) can be used to invoke the API. Obtain the Spreadsheet ID from the step 1. 
```
Remove the list numbering if there is only one item. Please apply below as well.

Author: Done

```
  Resource method: POST
  URL: http://localhost:8290/spreadsheet/read
```

  ```
    {
      "spreadsheetId":"1Ht0FWeKtKqBb1pEEzLcRMM8s5mktJdhivX3iaFXo-qQ",
      "range":"Employees!A1:C3"
    }
  ```

**Expected Response**: 
You should get the following response returned. 
```
  {
    "range": "Employees!A1:C3",
    "majorDimension": "ROWS",
    "values": [
      [
        "First Name",
        "Last Name",
        "Gender"
      ],
      [
        "John",
        "Doe",
        "Male"
      ],
      [
        "Leon",
        "Wins",
        "Female"
      ]
    ]
  }

```

### Spreadsheet Edit Operation

> 1. Invoke the SpreadsheetAPI with the following URL. Application such as [Postman](https://www.postman.com/) can be used to invoke the API. Obtain the Spreadsheet ID from the step 1. 

Reviewer 2: ```suggestion
1. Invoke the SpreadsheetAPI with the following URL. An application such as [Postman](https://www.postman.com/) can be used to invoke the API. Obtain the Spreadsheet ID from step 1. 
```

Author: Done

```
  Resource method: POST
  URL: http://localhost:8290/spreadsheet/edit
```

  ```
    {
      "spreadsheetId":"1Ht0FWeKtKqBb1pEEzLcRMM8s5mktJdhivX3iaFXo-qQ",
      "data": [
          {
            "values": [["Isuru","Uyanage","Female"],["Supun","Silva","Male"]],
            "range": "Employees!A6"
      }
      ]
    }
  ```

**Expected Response**: 
You should get the following response returned. 

```
  {
    "spreadsheetId": "1Ht0FWeKtKqBb1pEEzLcRMM8s5mktJdhivX3iaFXo-qQ",
    "totalUpdatedRows": 2,
    "totalUpdatedColumns": 3,
    "totalUpdatedCells": 6,
    "totalUpdatedSheets": 1,
    "responses": [
      {
        "spreadsheetId": "1Ht0FWeKtKqBb1pEEzLcRMM8s5mktJdhivX3iaFXo-qQ",
        "updatedRange": "Employees!A6:C7",
        "updatedRows": 2,
        "updatedColumns": 3,
        "updatedCells": 6
      }
    ]
  }
```

> The spreadsheet should be edited with in the above specified cell range. 

Reviewer 2: ```suggestion
The spreadsheet should be edited within the above specified cell range. 
```

Author: Done
